pkg/controller/wip/controller.go

```diff
@@ -0,0 +1,204 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package wip
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/client/cache"
+	k8sclientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
+	"k8s.io/kubernetes/pkg/util/runtime"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1"
+	servicecatalogclientset "github.com/kubernetes-incubator/service-catalog/pkg/client/clientset_generated/clientset"
+)
+
+// NewController returns a new Open Service Broker catalog
+// controller.
+func NewController(
+	kubeClient k8sclientset.Interface,
+	serviceCatalogClient servicecatalogclientset.Interface,
+	brokerInformer cache.SharedInformer,
+	serviceClassInformer cache.SharedInformer,
+	instanceInformer cache.SharedInformer,
+	bindingInformer cache.SharedInformer,
+) (Controller, error) {
+	controller := &controller{
+		kubeClient:           kubeClient,
+		serviceCatalogClient: serviceCatalogClient,
+		brokerInformer:       brokerInformer,
+		serviceClassInformer: serviceClassInformer,
+		instanceInformer:     instanceInformer,
+		bindingInformer:      bindingInformer,
+	}
+
+	brokerInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.brokerAdd,
+		UpdateFunc: controller.brokerUpdate,
+		DeleteFunc: controller.brokerDelete,
+	})
+
+	serviceClassInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.serviceClassAdd,
+		UpdateFunc: controller.serviceClassUpdate,
+		DeleteFunc: controller.serviceClassDelete,
+	})
+
+	instanceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.instanceAdd,
+		UpdateFunc: controller.instanceUpdate,
+		DeleteFunc: controller.instanceDelete,
+	})
+
+	bindingInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.bindingAdd,
+		UpdateFunc: controller.bindingUpdate,
+		DeleteFunc: controller.bindingDelete,
+	})
+
+	return controller, nil
+}
+
+// Controller describes a controller that backs the service catalog API for
+// Open Service Broker compliant Brokers.
+type Controller interface {
+	// Run runs the controller until the given stop channel can be read from.
+	Run(stopCh <-chan struct{})
+}
+
+type controller struct {
+	kubeClient           k8sclientset.Interface
+	serviceCatalogClient servicecatalogclientset.Interface
+	brokerInformer       cache.SharedInformer
+	serviceClassInformer cache.SharedInformer
+	instanceInformer     cache.SharedInformer
+	bindingInformer      cache.SharedInformer
+}
+
+var _ Controller = &controller{}
```

Comment:
This is unnecessary. The above `NewController` function will ensure that `controller` implements `Controller`